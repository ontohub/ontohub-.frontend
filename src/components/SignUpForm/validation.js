import { Client, userQuery } from "config/apollo";
import { get, isEmpty, filter, pick } from "lodash";
import debounce from "debounce-promise";

export const isUsernameAvailable = id =>
  Client.query({ query: userQuery, variables: { id: id } }).then(
    data => !get(data, "data.organizationalUnit", null)
  );

const isUsernameAvailableDebounced = debounce(isUsernameAvailable, 500, {
  leading: true
});

export const formHasError = errors =>
  !isEmpty(
    filter(pick(errors, ["email", "name", "password", "confirmPassword"]))
  );

export const addError = (condition, errors, field, error) => {
  if (condition) {
    if (!errors[field]) {
      errors[field] = [];
    }
    errors[field].push(error);
  }
};

export const validate = values => {
  return import("zxcvbn")
    .then(fn => {
      let errors = {};
      let scorePassword = password => fn(password).score;

      errors.passwordScore = scorePassword(values.password);

      addError(
        !values.password || values.password.length < 10,
        errors,
        "password",
        "Password must be at least 10 characters long"
      );
      addError(
        values.password !== values.confirmPassword,
        errors,
        "confirmPassword",
        "Passwords must match"
      );
      addError(
        !/^[^@]+@[^@]+$/.test(values.email),
        errors,
        "email",
        "Email is invalid"
      );
      addError(
        !values.name || values.name.length < 3,
        errors,
        "name",
        "Username must be at least 3 characters long"
      );
      addError(
        !/[a-z0-9][a-z0-9\-_]*[a-z0-9]/.test(values.name),
        errors,
        "name",
        "Username must start and end with a lower case letter " +
          "or number, and only contain lower case " +
          'letters, numbers, "-" and "_"'
      );
      return errors;
    })
    .then(errors => {
      /* istanbul ignore else */
      if (!errors.name) {
        return isUsernameAvailableDebounced(values.name).then(
          available => {
            addError(!available, errors, "name", "Username is already taken");
            /* istanbul ignore else */
            if (formHasError(errors)) {
              throw errors;
            } else {
              return errors;
            }
          },
          err => {
            addError(
              true,
              errors,
              "name",
              "Unable to check username availability. Please check your network connection."
            );
            throw errors;
          }
        );
      } else if (formHasError(errors)) {
        throw errors;
      } else {
        return errors;
      }
    });
};
