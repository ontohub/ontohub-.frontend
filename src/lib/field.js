import React from "react";
import { Form, Label } from "semantic-ui-react";

export const Field = ({
  bar,
  errors,
  fieldId,
  label,
  name,
  onBlur,
  onChange,
  touched,
  type,
  value
}) => (
  <Form.Field>
    <label htmlFor={fieldId}>{label}</label>
    <input
      id={fieldId}
      name={name}
      onBlur={onBlur}
      onChange={onChange}
      placeholder={label}
      type={type}
      value={value}
    />
    {bar || null}
    {!!touched &&
      !!errors && (
        <Label pointing color="red">
          {errors.map(e => <div key={e}>{e}</div>)}
        </Label>
      )}
  </Form.Field>
);
