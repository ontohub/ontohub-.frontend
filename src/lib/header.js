import React from "react";
import styled from "styled-components";

const Header = props => (
  <StyledHeader>
    <InnerHeader
      heightTransitionDuration={props.heightTransitionDuration}
      contentTransitionDuration={props.contentTransitionDuration}
    >
      {props.children}
    </InnerHeader>
  </StyledHeader>
);

const StyledHeader = styled.div`
  background-color: ${({ theme }) => theme.colors.dark};
  color: ${({ theme }) => theme.colors.light} !important;
  margin-bottom: 1em;
  margin-top: -100vh;
  padding-top: 100vh;
`;

const InnerHeader = styled.div`
  max-width: ${({ theme }) => theme.sizes.contentWidth};
  margin: auto;
  padding-top: ${({ theme }) => theme.sizes.headerPadding};
`;

export { Header };
