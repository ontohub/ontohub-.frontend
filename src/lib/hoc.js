import React from "react";

const HOC = ({ hoc = o => o, component, ...props }) =>
  React.createElement(hoc(component), props);

export { HOC };
