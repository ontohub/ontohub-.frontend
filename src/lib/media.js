import { css } from "styled-components";

// These sizes and key names are tailored to Semantic-UI
const sizes = {
  largeScreenOnly: { min: 1200 },
  computerOnly: { min: 992, max: 1199 },
  tabletOnly: { min: 768, max: 991 },
  mobileOnly: { max: 767 },

  largeScreen: { min: 1200 },
  computer: { min: 992 },
  tablet: { min: 768 },
  mobile: { min: 0 }
};

// use em in breakpoints to work properly cross-browser and support users
// changing their browsers font-size: https://zellwk.com/blog/media-query-units/
const emSize = pxSize => pxSize / 16;

// iterate through the sizes and create a media template
export const media = Object.keys(sizes).reduce((accumulator, label) => {
  const intervals = [];
  if (sizes[label].min) {
    intervals.push(`(min-width: ${emSize(sizes[label].min)}em)`);
  }
  if (sizes[label].max) {
    intervals.push(`(max-width: ${emSize(sizes[label].max)}em)`);
  }
  accumulator[label] = (...args) => css`
    @media ${intervals.join(" and ")} {
      ${css(...args)};
    }
  `;
  return accumulator;
}, {});
