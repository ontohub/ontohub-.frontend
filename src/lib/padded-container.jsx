import React from "react";
import styled from "styled-components";
import { Container } from "semantic-ui-react";

export const PaddedContainer = styled(Container)`
  padding-bottom: ${({ theme }) => theme.sizes.headerPadding};
`;
