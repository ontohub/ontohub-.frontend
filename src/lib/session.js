import React from "react";
import { ApolloConsumer, Mutation } from "react-apollo";
import gql from "graphql-tag";

const signOut = client => {
  localStorage.removeItem("auth-token");
  return client.resetStore();
};

const signIn = (client, token) => {
  if (token) {
    localStorage.setItem("auth-token", token);
    client.resetStore();
  }
};

const mutation = gql`
  mutation SignIn($username: String!, $password: String!) {
    signIn(username: $username, password: $password) {
      jwt
      __typename
    }
  }
`;

const SignIn = ({ username, password }) => (
  <Mutation mutation={mutation}>
    {(mutate, { data, loading, error, called }) => (
      <ApolloConsumer>
        {client => {
          if (!called && username && password) {
            mutate({ variables: { username, password } });
          } else if (called) {
            signIn(client, data.signIn.jwt);
          }
          return null;
        }}
      </ApolloConsumer>
    )}
  </Mutation>
);

const SignOut = () => (
  <ApolloConsumer>
    {client => {
      signOut(client);
      return null;
    }}
  </ApolloConsumer>
);

export { SignIn, SignOut, signIn, signOut, mutation };
