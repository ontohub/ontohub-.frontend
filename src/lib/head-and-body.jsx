import React, { Fragment } from "react";
import { HOC } from "lib/hoc";
import { Header } from "lib/header";

const HeadAndBody = ({ hoc, head: Head, body: Body, ...props }) => (
  <HOC
    hoc={hoc}
    component={props => (
      <Fragment>
        <Header>
          <Head {...props} />
        </Header>
        <Body {...props} />
      </Fragment>
    )}
    {...props}
  />
);

export { HeadAndBody };
