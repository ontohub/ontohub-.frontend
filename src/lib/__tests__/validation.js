import { setServerErrors } from "lib/validation";

describe("setServerErrors", () => {
  let setErrors;
  let setSubmitting;
  let onError;

  beforeEach(() => {
    setErrors = jest.fn();
    setSubmitting = jest.fn();
    onError = jest.fn();
    setServerErrors(setErrors, setSubmitting, onError)({
      message: `
        GraphQL error: Username not found
        GraphQL error: Password is missing
      `
    });
  });
  it("calls setErrors with the found errors", () => {
    expect(setErrors.mock.calls).toMatchSnapshot();
  });
  it("calls setSubmitting", () => {
    expect(setSubmitting.mock.calls).toEqual([[false]]);
  });
  it("calls onError", () => {
    expect(onError.mock.calls.length).toEqual(1);
  });

  describe("with an invalid error message", () => {
    beforeEach(() => {
      setErrors = jest.fn();
      setSubmitting = jest.fn();
      onError = jest.fn();
      setServerErrors(setErrors, setSubmitting, onError)({
        message: "This is an invalid error message"
      });
    });

    it("calls setErrors with the found errors", () => {
      expect(setErrors.mock.calls).toMatchSnapshot();
    });
  });
});
