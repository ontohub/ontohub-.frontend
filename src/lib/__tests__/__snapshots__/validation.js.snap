// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setServerErrors calls setErrors with the found errors 1`] = `
Array [
  Array [
    Object {
      "password": Array [
        "Password is missing",
      ],
      "username": Array [
        "Username not found",
      ],
    },
  ],
]
`;

exports[`setServerErrors with an invalid error message calls setErrors with the found errors 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;
